pkg/controller/atlasproject/ipaccess_list.go

```diff
@@ -42,18 +42,28 @@ func (r *AtlasProjectReconciler) ensureIPAccessList(ctx *workflow.Context, conne
 
 func validateIPAccessLists(ipAccessList []mdbv1.ProjectIPAccessList) error {
 	for _, list := range ipAccessList {
-		if list.DeleteAfterDate != "" {
-			_, err := timeutil.ParseISO8601(list.DeleteAfterDate)
-			if err != nil {
-				return err
-			}
+		if err := validateSingleIPAccessList(list); err != nil {
+			return err
 		}
-		// Go doesn't support XOR, but uses '!=' instead: https://stackoverflow.com/a/23025720/614239
-		onlyOneSpecified := isNotEmpty(list.AwsSecurityGroup) != isNotEmpty(list.CIDRBlock) != isNotEmpty(list.IPAddress)
-		if !onlyOneSpecified {
-			return errors.New("only one of the 'awsSecurityGroup', 'cidrBlock' or 'ipAddress' is required be specified")
+	}
+	return nil
+}
+
+// validateSingleIPAccessList performs validation of the IP access list. Note, that we intentionally don't validate
+// IP addresses or CIDR blocks - this will be done by Atlas. But we need to validate the timestamp as we use it to filter
+// active and expired ip access lists.
+func validateSingleIPAccessList(list mdbv1.ProjectIPAccessList) error {
+	if list.DeleteAfterDate != "" {
+		_, err := timeutil.ParseISO8601(list.DeleteAfterDate)
+		if err != nil {
+			return err
 		}
 	}
+	onlyOneSpecified := isNotEmpty(list.AwsSecurityGroup) != isNotEmpty(list.CIDRBlock) != isNotEmpty(list.IPAddress)
+	allSpecified := isNotEmpty(list.AwsSecurityGroup) && isNotEmpty(list.CIDRBlock) && isNotEmpty(list.IPAddress)
+	if !onlyOneSpecified || allSpecified {
+		return errors.New("only one of the 'awsSecurityGroup', 'cidrBlock' or 'ipAddress' is required be specified")
+	}
```

Comment:
[nit] May I suggest we replace this with a function? I know this is a one-time special thing, but this is really hard to read.

Maybe something like this: https://goplay.space/#889tv6pZ7l-

Reply:
sure!